### PR TITLE
AO3-4307 i18n admin_hidden_work_notification

### PR DESCRIPTION
**Pull request by zz9pzza**
## Issue

https://code.google.com/p/otwarchive/issues/detail?id=4316
https://otwarchive.atlassian.net/browse/AO3-4307

## Purpose

Make the email you receive when your work is hidden by an admin translatable.

### CODE REVIEW

File: config/locales/en.yml [hunk 1]
```diff
@@ -54,16 +54,28 @@ en:
...
+      subject: "[%{app_name}] Your work has been hidden by the Abuse Team"
+      hello: "Dear %{user},"
+      tos: "Terms of Service"
+      abuse: "contact our Abuse Team"
+      sincerely: "Sincerely,"
+      ao3_abuse: "AO3 Abuse"
+      part2: "We are investigating the matter and will contact you once the investigation has been completed. While the work is under investigation it has been hidden; only you will be able to access it."
+      html:
+        part1: "Your work <i>%{title}</i> has been reported as being in violation of the Archive of Our Own's %{tos}."
+        part3: "If you have any questions, please contact our Abuse Team (%{abuse})"
```
Comment (sarken) [inline config/locales/en.yml]: `part3` for the HTML email here says `please contact our Abuse Team (%{abuse})` but line 61 is `abuse: "contact our Abuse Team"`. 

So would the HTML email end up saying `please contact our Abuse Team (<a href="url for abuse">contact our Abuse Team</a>)` instead of the desired `please <a href="url for Abuse">contact our Abuse Team</a>`?

Comment (sarken) [inline config/locales/en.yml]: Oh, also! There should be a period at the end of the sentence for part3 of the HTML email, i.e. `part3: "If you have any questions, please %{abuse}."` (The text email's better without it, though.)

File: app/views/user_mailer/admin_hidden_work_notification.html.erb [hunk 2]
```diff
@@ -1,12 +1,11 @@
 <% content_for :message do %>
-    <p>Dear <%= @user.default_pseud.byline %>,</p>
+    <p><%= t('user_mailer.admin_hidden_work_notification.hello', user: @user.default_pseud.byline) %></p>
 
-    <p>Your work <i><%= style_link(@work.title.html_safe, work_url(@work)) %></i> has been reported as being in violation of the Archive of Our Own's <%= style_link("Terms of Service", tos_url) %>.</p>
+    <p><%= t('user_mailer.admin_hidden_work_notification.html.part1', title: style_link(@work.title.html_safe, work_url(@work)), tos: style_link('user_mailer.admin_hidden_work_notification.tos', tos_url).html_safe).html_safe%>
```
Comment (sarken) [inline app/views/user_mailer/admin_hidden_work_notification.html.erb]: There's a double `.html_safe` at the end of the line here -- is that needed to make the link work properly? 

Also, there's a space missing between the `.html_safe` and the `%>` at the very end: `.html_safe).html_safe%>`

File: app/mailers/user_mailer.rb [hunk 3]
```diff
@@ -363,11 +363,14 @@ def admin_deleted_work_notification(user, work)
...
     @work = Work.find_by_id(creation_id)
-
+    I18n.with_locale(Locale.find(@user.preference.preferred_locale).iso) do
     mail(
         to: @user.email,
-        subject: "[#{ArchiveConfig.APP_SHORT_NAME}] Your work has been hidden by the Abuse Team"
+        subject: t('user_mailer.admin_hidden_work_notification.subject', app_name: ArchiveConfig.APP_SHORT_NAME)
     )
+    end
+    ensure
```
Comment (sarken) [inline app/mailers/user_mailer.rb]: I think Hound is right that something wonky is up with the indenting here.

**Comment by sarken** [comment]
Can you track down the link and issue number for JIRA, please? This still has the old Google Code information in it, sob

File: app/views/user_mailer/admin_hidden_work_notification.html.erb (hunk 2 above)
Comment (zz9pzza) [inline app/views/user_mailer/admin_hidden_work_notification.html.erb]: The double .html_safe is needed as one is marking style_link('user_mailer.admin_hidden_work_notification.tos', tos_url) as safe while the other is marking the whole string.

**Comment by sarken** [comment]
👍

**Comment by sarken** [comment]
Question! Do you want to update the keys in this, too?

**Comment by zz9pzza** [comment]
I am updating it and putting my standard test which found an issue :)

"en.user_mailer.admin_hidden_work_notification.textl.part3"

File: app/views/user_mailer/admin_hidden_work_notification.text.erb [hunk 4]
```diff
@@ -1,12 +1,12 @@
...
+    <%= t('.hello', user: @user.default_pseud.byline) %>
 
-    Your work "<%= @work.title.html_safe %>" (<%= work_url(@work) %>) has been reported to the Abuse Team as a violation of the Archive of Our Own's Terms of Service (<%= root_url + "tos" %>).
+    <%= t('.text.part1', title: @work.title, tos_url: root_url + "tos" ) %> 
 
-    We are investigating the matter and will contact you once the investigation has been completed. While the work is under investigation it has been hidden; only you will be able to access it.
+    <%= t('.part2') %>
 
-    If you have any questions, please contact our Abuse Team (<%= root_url + "abuse_reports/new" %>).
+    <%= t('.text.part3', abuse: new_feedback_report_url ) %>
```
Comment (sarken) [inline app/views/user_mailer/admin_hidden_work_notification.text.erb]: There's an extra space before the `)` here. Also, this should be `new_abuse_report_url` -- feedback is for Support.

File: app/views/user_mailer/admin_hidden_work_notification.text.erb [hunk 5]
```diff
@@ -1,12 +1,12 @@
 <% content_for :message do %>
-    Dear <%= @user.default_pseud.byline %>,
+    <%= t('.hello', user: @user.default_pseud.byline) %>
 
-    Your work "<%= @work.title.html_safe %>" (<%= work_url(@work) %>) has been reported to the Abuse Team as a violation of the Archive of Our Own's Terms of Service (<%= root_url + "tos" %>).
+    <%= t('.text.part1', title: @work.title, tos_url: root_url + "tos" ) %>
```
Comment (sarken) [inline app/views/user_mailer/admin_hidden_work_notification.text.erb]: There's an extra space before the `)` here. Also, can this be `tos_url` instead of `root_url + "tos"`?

File: config/locales/en.yml [hunk 6]
```diff
@@ -54,16 +54,28 @@ en:
...
+      tos: "Terms of Service"
+      abuse: "contact our Abuse Team"
+      sincerely: "Sincerely,"
+      ao3_abuse: "AO3 Abuse"
+      part2: "We are investigating the matter and will contact you once the investigation has been completed. While the work is under investigation it has been hidden; only you will be able to access it."
+      html:
+        part1: "Your work <i>%{title}</i> has been reported as being in violation of the Archive of Our Own's %{tos}."
+        part3: "If you have any questions, please %{abuse}."
+      text:
+        part1: "Your work %{title} has been reported to the Abuse Team as a violation of the Archive of Our Own's Terms of Service (%{tos_url})"
```
Comment (sarken) [inline config/locales/en.yml]: There should be a period at the end of the sentence here, after the `)`. Also, in the existing email, the work title is in quotation marks and has the URL after at, like: `Your work "TITLE" (http://ao3.org/works/####)`.

ALSO, the phrasing of these two emails is inconsistent. /o\ I checked with Abuse, and could you change `part1` of the plain text email so the whole thing says: 

`Your work "%{title}" (%{work_url}) has been reported as being in violation of the Archive of Our Own's Terms of Service (%{tos_url}).`

That way, except for the necessary adjustments for links, the two emails actually give the same information with the same phrasing.

File: config/locales/en.yml [hunk 7]
```diff
@@ -54,16 +54,28 @@ en:
...
+        part3: "If you have any questions, please contact our Abuse Team (%{abuse})"
```
Comment (sarken) [inline config/locales/en.yml]: This is also missing the period at the end of the sentence, after the `)`.

File: app/views/user_mailer/admin_hidden_work_notification.html.erb [hunk 8]
```diff
@@ -1,12 +1,11 @@
 <% content_for :message do %>
-    <p>Dear <%= @user.default_pseud.byline %>,</p>
+    <p><%= t('.hello', user: @user.default_pseud.byline) %></p>
 
-    <p>Your work <i><%= style_link(@work.title.html_safe, work_url(@work)) %></i> has been reported as being in violation of the Archive of Our Own's <%= style_link("Terms of Service", tos_url) %>.</p>
+    <p><%= t('.html.part1', title: style_link(@work.title.html_safe, work_url(@work)), tos: style_link('.tos', tos_url).html_safe).html_safe %>
```
Comment (sarken) [inline app/views/user_mailer/admin_hidden_work_notification.html.erb]: It looks like the closing `</p>` tag got deleted at the end of this line.

File: config/locales/en.yml [hunk 9]
```diff
@@ -54,16 +54,28 @@ en:
...
+      abuse: "contact our Abuse Team"
+      sincerely: "Sincerely,"
+      ao3_abuse: "AO3 Abuse"
+      part2: "We are investigating the matter and will contact you once the investigation has been completed. While the work is under investigation it has been hidden; only you will be able to access it."
+      html:
+        part1: "Your work <i>%{title}</i> has been reported as being in violation of the Archive of Our Own's %{tos}."
+        part3: "If you have any questions, please %{abuse}."
+      text:
+        part1: "Your work %{title} has been been reported as being in violation of the Archive of Our Own's Terms of Service (%{tos_url})."
+        part3: "If you have any questions, please contact our Abuse Team (%{abuse})"
```
Comment (sarken) [inline config/locales/en.yml]: Needs a period after the closing parenthesis in `(%{abuse})`. 

Other than that, 👍

File: config/locales/en.yml [hunk 10]
```diff
@@ -53,6 +53,20 @@ en:
...
+    admin_hidden_work_notification:
+      subject: "[%{app_name}] Your work has been hidden by the Abuse Team"
+      hello: "Dear %{user},"
+      tos: "Terms of Service"
+      abuse: "contact our Abuse Team"
+      sincerely: "Sincerely,"
+      ao3_abuse: "AO3 Abuse"
+      part2: "We are investigating the matter and will contact you once the investigation has been completed. While the work is under investigation it has been hidden; only you will be able to access it."
+      html:
+        part1: "Your work <i>%{title}</i> has been reported as being in violation of the Archive of Our Own's %{tos}."
```
Comment (sarken) [inline config/locales/en.yml]: Can we reduce the amount of HTML the translators have to work with by using the same `<% pretty_title = "<i>#{style_bold(@work.title.html_safe)}</i>".html_safe %>` approach here that we used in [delete_work_notification](https://github.com/otwcode/otwarchive/blob/master/app/views/user_mailer/delete_work_notification.text.erb)?

File: app/views/user_mailer/admin_hidden_work_notification.html.erb [hunk 11]
```diff
@@ -1,12 +1,12 @@
 <% content_for :message do %>
-    <p>Dear <%= @user.default_pseud.byline %>,</p>
+<% @pretty_title = "<i>#{style_bold(@work.title.html_safe)}</i>".html_safe %>
```
Comment (sarken) [inline app/views/user_mailer/admin_hidden_work_notification.html.erb]: The work title is a link in the current version, so I think this should be more like: 
```
<% pretty_title = "<i>#{style_link(@work.title.html_safe, work_url(@work))}</i>".html_safe %>
```

Also,  just plain `pretty_title` should work fine. I think the reason it complained before was a mistake in the en.yml part, which I will comment on down there.

File: config/locales/en.yml [hunk 12]
```diff
@@ -51,8 +51,22 @@ en:
...
+    admin_hidden_work_notification:
+      subject: "[%{app_name}] Your work has been hidden by the Abuse Team"
+      hello: "Dear %{user},"
+      tos: "Terms of Service"
+      abuse: "contact our Abuse Team"
+      sincerely: "Sincerely,"
+      ao3_abuse: "AO3 Abuse"
+      part2: "We are investigating the matter and will contact you once the investigation has been completed. While the work is under investigation it has been hidden; only you will be able to access it."
+      html:
+        part1: "Your work <i>%{pretty_title}</i> has been reported as being in violation of the Archive of Our Own's %{tos}."
```
Comment (sarken) [inline config/locales/en.yml]: `%{pretty_title}` is just supposed to be `%{title}`, since that's the name of the key. Also, the `<i>` and `</i>` should be removed here.

Comment (zz9pzza) [inline config/locales/en.yml]: Thank you, I was being stupid there.

**Comment by sarken** [comment]
Marking Needs Coder Action because this has merge conflicts

File: config/locales/en.yml [hunk 13]
```diff
@@ -5,7 +5,7 @@ en:
     blacklist:
       emails_found:
         one: "one email found"
-        other: "%{count} emails found"
```
Comment (sarken) [inline config/locales/en.yml]: There's a test failure related to this line.

File: config/locales/en.yml [hunk 14]
```diff
@@ -152,6 +152,20 @@ en:
           friend at %{invitation_page}.
           Cheers,
           %{archive_name}
+    admin_hidden_work_notification:
+      subject: "[%{app_name}] Your work has been hidden by the Abuse Team"
+      hello: "Dear %{user},"
+      tos: "Terms of Service"
+      abuse: "contact our Abuse Team"
```
Comment (sarken) [inline config/locales/en.yml]: Can I suggest clearer, more distinct keys for lines 159 and 161, e.g. `contact_abuse` for this and `signature` for 161?

File: config/locales/en.yml [hunk 15]
```diff
@@ -197,6 +211,11 @@ en:
     change_email:
       subject: "[%{app_name}] Email changed"
       part1: "%{login}, the email associated with your account has been changed to %{email}"
+    reset_password:
```
Comment (sarken) [inline config/locales/en.yml]: I don't think this belongs in this PR... it seems to be the old version of the password reset email, which we changed when we moved to Devise.

File: app/views/user_mailer/admin_hidden_work_notification.html.erb [hunk 16]
```diff
@@ -1,12 +1,12 @@
 <% content_for :message do %>
```
Comment (sarken) [inline app/views/user_mailer/admin_hidden_work_notification.html.erb]: The indenting here is a bit wonky. I think it should be:
```
<% content_for :message do %>
  <% pretty_title = "<i>#{style_link(@work.title.html_safe, work_url(@work))}</i>".html_safe %>
  <p><%= t(".hello", user: @user.default_pseud.byline) %></p>

  <p><%= t(".html.part1", title: pretty_title, tos: style_link(t(".tos"), tos_url).html_safe).html_safe %></p>

  <p><%= t(".part2") %></p>

  <p><%= t(".html.part3", abuse: abuse_link(t(".contact_abuse"))).html_safe %></p>
  <p><%= t(".sincerely") %></p>
  <p><%= style_bold(t(".signature")) %></p>
<% end %>
```

File: config/locales/en.yml [hunk 17]
```diff
@@ -152,6 +152,20 @@ en:
...
+      contact_abuse: "contact our Abuse Team"
+      sincerely: "Sincerely,"
+      signature: "AO3 Abuse"
+      part2: "We are investigating the matter and will contact you once the investigation has been completed. While the work is under investigation it has been hidden; only you will be able to access it."
+      html:
+        part1: "Your work %{title} has been reported as being in violation of the Archive of Our Own's %{tos}."
+        part3: "If you have any questions, please %{abuse}."
+      text:
+        part1: "Your work %{title} has been been reported as being in violation of the Archive of Our Own's Terms of Service (%{tos_url})."
+        part3: "If you have any questions, please %{abuse}"
```
Comment (sarken) [inline config/locales/en.yml]: I think this should be `part3: "If you have any questions, please contact our Abuse Team (%{abuse})."`. Otherwise, the email will say, "If you have any questions, please https://archiveofourown.org/abuse_reports/new."

File: features/admins/admin_works.feature [hunk 18]
```diff
@@ -24,6 +24,7 @@ Feature: Admin Actions for Works, Comments, Series, Bookmarks
       And logged in users should not see the hidden work "ToS Violation" by "regular_user"
       And "regular_user" should see their work "ToS Violation" is hidden
       And 1 email should be delivered
+      And the email should not contain "translation missing"
```
Comment (sarken) [inline features/admins/admin_works.feature:27]: We'll need some specs for this in spec/mailers/user_mailer_spec.rb as well.

File: spec/mailers/user_mailer_spec.rb [hunk 19]
```diff
@@ -381,4 +381,40 @@
       end
     end
   end
+
+  describe "admin_hidden_notification" do
+    title = Faker::Book.title
+    let(:author) { create(:user) }
+    let(:work) { create(:work, title: title, authors: [author.pseuds.first]) }
+
+
```
[the remark this answers is not in the document]
Comment (redsummernight) [reply spec/mailers/user_mailer_spec.rb]: Can you fix the Hound comments in this file?

File: spec/mailers/user_mailer_spec.rb [hunk 20]
```diff
@@ -381,4 +381,40 @@
...
+    subject(:email) { UserMailer.admin_hidden_work_notification(work.id,  author.id).deliver }
```
[the remark this answers is not in the document]
Comment (redsummernight) [reply spec/mailers/user_mailer_spec.rb:390]: Can you fix the Hound comment in this file?

File: app/views/user_mailer/admin_hidden_work_notification.html.erb [hunk 21]
```diff
@@ -1,12 +1,12 @@
...
-    <p>Dear <%= @user.default_pseud.byline %>,</p>
+  <% pretty_title = "<i>#{style_link(@work.title.html_safe, work_url(@work))}</i>".html_safe %>
```
Comment (sarken) [inline app/views/user_mailer/admin_hidden_work_notification.html.erb:2]: I need someone else to check me on this, but I'm not sure if we need the `html_safe` on `@work.title`, since we've overridden the [title method](https://github.com/otwcode/otwarchive/blob/17307a8a6294d2af7fab882ea5aa14eb786d51af/app/models/work.rb#L74-L77) to make it always use `html_safe`. 

I tried taking the `html_safe` out and testing with a title that had an ampersand. It looked okay:

```<p>Your work <i><a style="color:#990000" href="http://www.example.com/works/7">Modern & Old-fashioned</a></i> has been reported as being in violation of the Archive of Our Own's <a style="color:#990000" href="http://www.example.com/tos">Terms of Service</a>.</p>```

But it also looked okay when, in addition to removing the `html_safe` call on the title, I removed
- the override on the `title` method
- the `html_safe` at the end of this line
- the `html_safe` on `tos: style_link(t(".tos"), tos_url).html_safe`

It only broke when I finally removed the `html_safe` at the very end of line 5.

tl;dr: we might be being overzealous with our `html_safe` -- it's possible we don't need any on line 2 and we only need the final one on line 5.

File: spec/mailers/user_mailer_spec.rb [hunk 22]
```diff
@@ -381,4 +381,36 @@
...
+    let(:work) { create(:work, title: title, authors: [author.pseuds.first]) }
+
+    subject(:email) { UserMailer.admin_hidden_work_notification(work.id,  author.id).deliver }
+    # Test the headers
+    it_behaves_like "an email with a valid sender"
+
+    it "has the correct subject line" do
+      expect(email.subject).to eq("[#{ArchiveConfig.APP_SHORT_NAME}] Your work has been hidden by the Abuse Team")
+    end
+    # Test both body contents
```
Comment (sarken) [inline spec/mailers/user_mailer_spec.rb:397]: Could you add a blank line between this and line 396?

File: spec/mailers/user_mailer_spec.rb [hunk 23]
```diff
@@ -381,4 +381,36 @@
...
+      it "has the correct content" do
+        expect(email).to have_html_part_content("While the work is under investigation it has been hidden")
+        expect(email).to have_html_part_content("contact our Abuse Team</a>")
+      end
+    end
+
+    describe "text version" do
+      it "has the correct content" do
+        expect(email).to have_text_part_content("While the work is under investigation it has been hidden")
+        expect(email).to have_text_part_content("contact our Abuse Team")
```
Comment (sarken) [inline spec/mailers/user_mailer_spec.rb:412]: Maybe include the opening parenthesis in the check here, i.e. `contact our Abuse Team (`?

File: app/views/user_mailer/admin_hidden_work_notification.html.erb (hunk 21 above)
Comment (sarken) [reply app/views/user_mailer/admin_hidden_work_notification.html.erb:2]: I'm going to forget if I don't comment now: we should use the style_creation_link helper. https://github.com/otwcode/otwarchive/blob/8d1456a2b876e5d743c6e71748f8b0cb87f6eb1d/app/helpers/mailer_helper.rb#L12-L16
